src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
This flag is updated by progress_engine->lock/unlock_ack callback, and waited in lock or unlock call. 

My understanding is that such update/wait can be performed by different threads, thus being dangerous with compiler optimization.